### PR TITLE
fix(agents-api): init nlp pipeline text-search

### PR DESCRIPTION
**Pull request by Vedantsahai18**
### **PR Type**
Enhancement, Tests

___

### **Description**
- Enhanced NLP pipeline for text search with improved keyword extraction.

- Refactored query-building logic for efficiency and clarity.

- Added tests for embedding generation and text-to-query conversion.

- Introduced utility for generating vectors with specific cosine similarity.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>nlp.py</strong><dd><code>Enhanced NLP pipeline and query-building logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

agents-api/agents_api/common/nlp.py

<li>Improved stopword filtering for noun chunks and spans.<br> <li> Refactored query-building logic for efficiency and clarity.<br> <li> Renamed and optimized functions for text-to-query conversion.<br> <li> Increased caching capacity for query-related functions.

</details>

  </td>
  <td><a href="https://github.com/julep-ai/julep/pull/1045/files#diff-66deb57e966d8a3ac8b61fe3240fff24ef1d0aa2f31e9303c88df372245c3be0">+61/-64</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>search_docs_by_text.py</strong><dd><code>Preprocess raw text queries for search</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

agents-api/agents_api/queries/docs/search_docs_by_text.py

- Added preprocessing step for raw text queries.

</details>

  </td>
  <td><a href="https://github.com/julep-ai/julep/pull/1045/files#diff-d8e9417b5496c44804e81c8ed8f8dd289ad387e5a7595d786987b2c97adb050c">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>utils.py</strong><dd><code>Added vector generation utility for testing</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

agents-api/tests/utils.py

<li>Introduced utility to generate vectors with specific cosine <br>similarity.<br> <li> Added mathematical handling for edge cases in vector generation.

</details>

  </td>
  <td><a href="https://github.com/julep-ai/julep/pull/1045/files#diff-f57e08363e297b9a26b89a750f00ff1dbcbfdba9189591c528f35691d58a7432">+47/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>fixtures.py</strong><dd><code>Added test embeddings with confidence levels</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

agents-api/tests/fixtures.py

<li>Added embeddings with varying confidence levels for testing.<br> <li> Inserted test data into the database for embeddings.

</details>

  </td>
  <td><a href="https://github.com/julep-ai/julep/pull/1045/files#diff-e6fdaaf046732cdbe189837561b0795f04f1235a6752d746f750330db11ca509">+53/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>test_docs_queries.py</strong><dd><code>Added tests for document search and text-to-query utility</code></dd></summary>
<hr>

agents-api/tests/test_docs_queries.py

<li>Added tests for searching documents with technical terms.<br> <li> Verified document retrieval based on query relevance.<br> <li> Added tests for text-to-query utility function.

</details>

  </td>
  <td><a href="https://github.com/julep-ai/julep/pull/1045/files#diff-36539fadcf9fdd23428ea921b45082c021d6816dddf931a806b9828fa43d1fb1">+171/-0</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

----

> [!IMPORTANT]
> Enhances NLP pipeline for text search, refactors query logic, and adds tests for embedding and text-to-query conversion.
> 
>   - **Enhancements**:
>     - Improved NLP pipeline in `nlp.py` with better stopword filtering and optimized query-building logic.
>     - Added preprocessing step for raw text queries in `search_docs_by_text.py`.
>     - Introduced utility in `utils.py` for generating vectors with specific cosine similarity.
>   - **Tests**:
>     - Added tests in `test_docs_queries.py` for document search and text-to-query conversion.
>     - Added embeddings with varying confidence levels in `fixtures.py` for testing.
>   - **Refactoring**:
>     - Refactored query-building logic in `nlp.py` for efficiency and clarity.
>     - Renamed functions for text-to-query conversion in `nlp.py`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=julep-ai%2Fjulep&utm_source=github&utm_medium=referral)<sup> for 27ed1f4a2944f7afb5e5a1ab483693973187297e. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: agents-api/agents_api/common/nlp.py [hunk 1]
```diff
@@ -180,35 +180,162 @@ def union(u: str, v: str) -> None:
...
+#             pattern = build_query_pattern(len(group), n)
+#             clause = pattern.format(*sorted_group)
+#             clauses.append(clause)
+
+#     return " OR ".join(clauses)
+
+
+# @lru_cache(maxsize=100)
+# def paragraph_to_custom_queries(
+#     paragraph: str, top_n: int = 10, proximity_n: int = 10, min_keywords: int = 1
```
Comment (creatorrr) [inline agents-api/agents_api/common/nlp.py:210]: We can remove proximity_n from the arguments and the function

Comment (Vedantsahai18) [reply agents-api/agents_api/common/nlp.py]: Done

File: agents-api/agents_api/common/nlp.py [hunk 2]
```diff
@@ -180,35 +180,162 @@ def union(u: str, v: str) -> None:
...
-    return " OR ".join(clauses)
-
-
-@lru_cache(maxsize=100)
-def paragraph_to_custom_queries(
+# def build_query_pattern(group_size: int, n: int) -> str:
+#     """Cache query patterns for common group sizes."""
+#     if group_size == 1:
+#         return '"{}"'
+#     return f"NEAR/{n}(" + " ".join('"{}"' for _ in range(group_size)) + ")"
```
[the remark this answers is not in the document]
Comment (Vedantsahai18) [reply agents-api/agents_api/common/nlp.py]: Removed
